PCF-ADR-0004-codeflare-test-strategy.md

```diff
@@ -0,0 +1,392 @@
+             - emptyDir:
+                 medium: Memory
+               name: dshm
+       priority: 0
+       priorityslope: 0
+       replicas: 1
+     - allocated: 0
+       generictemplate:
+         apiVersion: v1
+         kind: Service
+         metadata:
+           name: test-job
+         spec:
+           clusterIP: None
+           ports:
+             - port: 29500
+               protocol: TCP
+               targetPort: 29500
+           publishNotReadyAddresses: true
+           selector:
+             appwrapper.mcad.ibm.com: test-job
+           sessionAffinity: None
+           type: ClusterIP
+ schedulingSpec:
+   requeuing:
+     growthType: exponential
+     maxNumRequeuings: 0
+     maxTimeInSeconds: 0
+     numRequeuings: 0
+     timeInSeconds: 300
+```
+
+3. Wait until the test job has completed
+4. Assert the job status is successful
+5. Delete the test AppWrapper resource
+6. Assert all the resources have been successfully freed
+
+#### OLM Installation / Upgrade Test Cases
+
+##### Setup
+
+The following steps must be executed once, with a **cluster admin role**, as prerequisites to running the OLM installation / upgrade test cases:
+
+1. Provision a test Kubernetes cluster (resp. an OpenShift cluster)
+2. For Kubernetes cluster only: install OLM by executing the following commands:
+```
+$ kubectl apply -f https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.24.0/crds.yaml
+$ kubectl apply -f https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.24.0/olm.yaml
+```
+3. Build the CodeFlare operator OLM bundle:
+  * Clone the CodeFlare operator source code repository
+  * Checkout the branch to be tested, e.g., the PR feature branch, i.e. `git clone --branch mytag0.1 --depth 1 https://example.com/my/repo.git`
+  * Build the CodeFlare operator container image at HEAD revision
+  * Push the CodeFlare operator container image into a test container image registry (resp. the OpenShift cluster internal image registry)
+  * Build the OLM bundle image
+  * Push the OLM bundle image into a test container image registry (resp. the OpenShift cluster internal image registry)
+
+4. Build the OLM index image:
+  * Build the OLM index/catalog image with OPM from the latest Operator Hub community catalog (resp. OpenShift catalog), e.g.:
+```
+$ mkdir catalog
+$ opm render registry.redhat.io/redhat/redhat-operator-index:<OCP_VERSION> -o yaml > catalog/bundles.yaml
+$ opm render $BUNDLE_IMAGE > catalog/codeflare-operator-bundle.yaml
+$ sedtemp=$(mktemp sed-template-XXX.sed)
+$ cat << EOF > ${sedtemp}
+/- name: codeflare-operator.v${PREVIOUS_VERSION}/ {
+  p;
+  n;
+  /  replaces:/ {
+    p;
+    n;
+    /name: alpha$/ {
+      i- name: codeflare-operator.v${PREVIOUS_VERSION}
+      i\ \ replaces: codeflare-operator.v${PREVIOUS_VERSION}
+      p;
+      d;
+    }
+  }
+}
+p;
+EOF
+$ sed -i -n -f ${sedtemp} ${CATALOG_DIR}/bundles.yaml
+$ rm -f ${sedtemp}
```

Comment:
It can be replaced with the simpler sed command that does not try to preserve channel entries ordering.